pkg/storage/gcsstorage.go

```diff
@@ -0,0 +1,233 @@
+package storage
+
+// Fork from Thanos GCS Bucket support to reuse configuration options
+// Licensed under the Apache License 2.0.
+// https://github.com/thanos-io/thanos/blob/main/pkg/objstore/gcs/gcs.go
```

Comment:
Is this all we need for attribution? Should we reference a commit it came from, or any specific authors?

Reply:
Any specific changes you made to the fork worth pointing out, to help us poor reviewers?

Reply:
Hahaha! I see that this was the last thing you noticed as well. It's the same for all the bucket storage implementations. Based on what I have read concerning Apache 2, this is all that is required, but I would _love_ a confirmation there. 

Mainly, the fork was to ensure 1-1 with the Thanos Bucket Configuration for all of these bucket types. So, any functionality which could be configured in the Thanos format was forked and used. However, their "Bucket" API isn't the same as ours at all, it's designed around specific thanos use-cases (like recursively diving into sub-directories), so most of the implementation of the `storage.Storage` interface is "guided" by their implementation, but it's mostly different. The Azure implementations are probably the closest since the blob storage APIs for Azure are a complete disaster.